src/mrtool/core/utils.py

```
@@ -426,3 +150,60 @@ def proj_simplex(x: np.ndarray) -> np.ndarray:
     lam = v[rho]
 
     return np.maximum(x + lam, 0.0)
+
+
+def sample_knots(num_knots: int = 5,
+                 sample_width: float = 0.05,
+                 sample_size: int = 1) -> np.ndarray:
+    """
+    Sample knots
+
+    Parameters
+    ----------
+    num_knots : int, optional
+        Total number of knots, include the boundary knots, must be greater or
+        equal than 2. Default to be 5.
+    sample_width : float, optional
+        Sample interval for each knot. For example, if ``num_knots=3``, we have
+        one interior knots and the average position is 0.5.
+        If ``sample_width=0.1``, we sample the interior knot between,
+        ``[0.5 - 0.1, 0.5 + 0.1]``. Default to be 0.05.
+    sample_size : int, optional
+        Number of samples, must be greater or equal to 1. Default to be 1.
+
+    Returns
+    -------
+    ndarray
+        Knots samples matrix, with shape ``(sample_size, num_knots)``.
+    """
+    if num_knots < 2:
+        raise ValueError("Number of knots must be greater or equal than 2.")
+    if sample_width < 0.0 or sample_width > 1.0:
+        raise ValueError("sample_width need to be between 0 and 1.")
+    if sample_size < 1:
+        raise ValueError("sample_size at least need to be 1.")
+
+    knots = np.linspace(0, 1, int(num_knots))
+    inner_knots = knots[1:-1]
+    if inner_knots.size == 0:
+        return np.tile(knots, (sample_size, 1))
+
+    inner_knots_bounds = np.vstack([
+        np.minimum(1.0, np.maximum(0.0, inner_knots - sample_width)),
+        np.minimum(1.0, np.maximum(0.0, inner_knots + sample_width))
```

I like these two lines. It can avoid problems caused by unreasonable sample intervals. But I'm wondering if we need to raise an error if an unreasonable `sample_width` is provided, for example, `sample_width` > 0.25 when `num_knots` = 5

---

Thanks @jiaweih ! I think for now we could simplify the requirement of `sample_width` to be only between 0 and 1. So that it is easier for the user :) We could treat this as the intended behavior.